### PR TITLE
fix: encoder port validation, lint

### LINKED ISSUE #252: Encoder error handling is broken in multiple ways.
## Bug Description
Specifically regarding port validation, there are some logic errors in how encoders are created. I believe it allows you to construct encoders on ports that shouldn't be allowed and explicitly disallows you to create encoders in valid positions.

### PR DESCRIPTION
**Pull request by Tropix126**
## Describe the changes this PR makes. Why should it be merged?
This PR fixes some misconceptions we had about encoder ports. VEX article summarizes it well:

image
 
Also resolves some leftover lints from the nightly update a while back.

Closes #252

### CODE REVIEW

**Comment by zabackary** [comment]
In terms of API design, since you all seem to prefer infallible device constructors, do you think it would be a good idea for these constructors to panic? 

To quote the Rust book: 

> Functions often have contracts: their behavior is only guaranteed if the inputs meet particular requirements. Panicking when the contract is violated makes sense because a contract violation always indicates a caller-side bug, and it’s not a kind of error you want the calling code to have to explicitly handle.

**Comment by Gavin-Niederman** [comment]
> In terms of API design, since you all seem to prefer infallible device constructors, do you think it would be a good idea for these constructors to panic?
> 
> _edit: by infallible, I mean "doesn't return `Result`"_

In the case of this constructor, yes I think panicking on an invalid port pair is reasonable. I don't think that panicking on a device error is reasonable because that's impossible to recover from. So In the case of other constructors I think it is better to silently fail or return the error.

**Comment by Tropix126** [comment]
> In terms of API design, since you all seem to prefer infallible device constructors, do you think it would be a good idea for these constructors to panic? 
> 
> _edit: by infallible, I mean "doesn't return `Result`"_
> 
> To quote the Rust book: 
> 
> > Functions often have contracts: their behavior is only guaranteed if the inputs meet particular requirements. Panicking when the contract is violated makes sense because a contract violation always indicates a caller-side bug, and it’s not a kind of error you want the calling code to have to explicitly handle.

I agree with this notion, and we'll probably change this in the next breaking update.

**Comment by zabackary** [comment]
> in the next breaking update.

...which will probably be the next one, since almost every release recently has been either breaking or a docs.rs fix.

**Comment by Tropix126** [comment]
> > in the next breaking update.
> 
> ...which will probably be the next one, since almost every release recently has been either breaking or a docs.rs fix.

There will be at least one intermediate update following 0.5.0 to address some issues.

File: packages/vexide-devices/src/adi/encoder.rs [hunk 1]
```diff
@@ -114,16 +116,15 @@ impl AdiEncoder {
...
-            top_port.index() % 2 != 0,
-            BadTopPortSnafu {
-                port: top_port.number()
-            }
-        );
-        // Bottom must be directly next to top on the higher port index.
+
+        // Top and bottom must be some combination of (AB, CD, EF, GH) or (BA, CD, FE, HG)
+        let top_is_even = top_number % 2 == 0;
         ensure!(
```
Comment (Gavin-Niederman) [inline packages/vexide-devices/src/adi/encoder.rs:122]: Not going to block merging for this, but in the future this should probably be a panic.

**Review by Gavin-Niederman — APPROVED** [review]
LGTM

**Pull request merged**